### PR TITLE
Marge Component system to main

### PR DESCRIPTION
**Pull request by DelCherub**
This involves:
-Component.h, component base class with Unity-like virtual methods for Start, Update, PhysicsUpdate and OnCollision methods.
-Added entt.hpp, external entity component system. We do not plan on using the ECS of this, but DO intend to use the signal functionality.
-Changes to GameObject.h to have methods which call all components at an appropriate time: when GameObject is Updated, calls Update() function on all components etc. Some of the new functionality uses shared pointers, some uses raw pointers purely because the codebase uses raw pointers and there is too much to change.

NOT IMPLEMENTED IN THIS BRANCH
- The methods of the GameObject are not called at any point, and need to be added in whichever class should call them. GameWorld must call the Update() function of GameObjects when looping though them, must also call Start() once all objects have been initialised.

### CODE REVIEW

**Review by St4rman — COMMENTED** [review]
lgtm !!!! :]

**Review by St4rman — APPROVED** [review]
lgtm part 2

**Review by Joerseven — CHANGES_REQUESTED** [review]
Still thinking about this, discussed as a group

File: Replicated/Component.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	class Component
+	{
+	public:
+		Component() { };
+		~Component() {};
+		Component(GameObject*  gameObject) { this->gameObject = gameObject;}
+
+		//all virtual, not the best approach but works
+		virtual void Update(float dt) {};
+		virtual void PhysicsUpdate(float dt) {};
```
Comment (Joerseven) [inline Replicated/Component.h]: Rename fixed time over dt.

File: CSC8503CoreClasses/GameObject.h [hunk 2]
```diff
@@ -1,19 +1,43 @@
...
+			for (Component* component : components)component->OnCollisionEnd(otherObject);
+		}
+
+		virtual void OnCollisionStay(const GameObject* otherObject) {
+			for (Component* component : components)component->OnCollisionStay(otherObject);
+		}
+
+		void UpdateAllComponents(float dt) { for (Component* component : components)component->Update(dt); }
+
+		void PhysicsUpdateAllComponents(float dt) { for (Component*& component : components)component->PhysicsUpdate(dt); }
```
Comment (Joerseven) [inline CSC8503CoreClasses/GameObject.h]: Big memes

File: CSC8503CoreClasses/GameObject.h [hunk 3]
```diff
@@ -74,8 +90,26 @@ namespace NCL::CSC8503 {
 			return worldID;
 		}
 
-		void DrawCollision();
+		//returns true if component found, false if not
+		template <typename T>
+		bool TryGetComponent(T*& returnPointer) {
+			for (Component* component : components) {
+				T* typeCast = dynamic_cast<std::shared_ptr<T>>(component);
```
Comment (Joerseven) [inline CSC8503CoreClasses/GameObject.h]: Normal pointer here.

**Review by Joerseven — DISMISSED** [review]
Suggestions put in!

**Review by Odycious — APPROVED** [review]

**Pull request merged**